roles/micado_master/build/files/terraform/terraform

```diff
@@ -0,0 +1,11 @@
+#/bin/bash
+
+cd /var/lib/micado/terraform/submitter/
+
+FILE=preprocess.py
+if [ -f "$FILE" ]; then
```

Comment:
Does this file need to be edited each time a new pre-processing script is added? Can the scripts be picked from a directory rather than hardcoded?

Reply:
I was under the impression of keeping the preprocess.py and postprocess.py as default names for all the pre and post processing scripts.

Reply:
What if there are multiple preprocessing scripts?

Reply:
Can you please elaborate on that. Now also we have several files needed for preprocessing which are all being called from preprocess.py.

Reply:
Imagine we have a use case that needs a token for authenticating with AWS. I might have a pre-processing script that fetches the token from an endpoint and then writes it to the translated tf.json file.

I'll call it aws-token.py
```python3
import glob, json, requests

tf_files = glob.glob("*.tf.json")
for tf_file in tf_files:
    with open(tf_file) as tf:
        json_data = json.load(tf)

    if json_data["provider"].get("aws") and SOME_OTHER_CONDITIONS:        
        aws_token = requests.get(AWS_TOKEN_ENDPOINT).json()["token"]
        json_data["provider"]["aws"]["token"] = aws_token
```

I'd like to simply drop the aws-token.py file into the pre-processing script directory and it should be run accordingly

Reply:
But currently the preprocessing script is copied to terraform directory using shutil-copy in adaptor. So the name can be made preprocess.py there. But then there it should take files with any name. Or are we mounting these files also directly?

Reply:
I think they should be mounted in another way (micado-manifest probably?), otherwise the TerraformAdaptor has to be modified each time a new script is added to MiCADO.

Reply:
It is good to that but I need to be more clarity on this. If like you said if the pre-processing script files are having different names for different cloud, we can mount the files/directory as needed. But the problem if like in the case of egi we have 3 python files needed for preprocess (preprocess.py, configure.py, pyGetScopedToken.py) where preprocess.py acts as the main which call the other scripts if and when needed. If it was only 1 file in the wrapper script I can check whether there is any file in a particular directly and execute it but in the case where there are more than one file I am not sure what you had in mind.

Reply:
Good question. Maybe all the scripts in the root pre-processing directory are executed by the wrapper? These main scripts can have helper scripts located in sub-directories? Otherwise a naming convention?

@smith4 : what is your opinion on this?

Reply:
If we consider that for any cloud, we should be able to add any number of py files without name conflicts among them, putting the py files into a directory structure could be a solution. We can choose a fix py filename as entry point and the directory structure is searched for these py files. e.g. entry point = "preprocess.py".
- EGI - preprocess.py, configure.py, pyGetScopedToken.py
- AWS - preprocess.py, anyhelper.py, additionallib.py
In the terraform wrapper, we can implement the directory scanning to find "preprocess.py" files. Or, if we name the directories after the cloud provider name then searching is not needed, just checking the existence of file "<provider_name>/preprocess.py and then run it.
Let me know your thoughts on this proposal.

Reply:
Can it happen that the worker nodes in MiCADO are deployed under more than one EGI account? I mean that multiple tokens should be handled at the same time?

Reply:
Based on my understanding we are currently handling only one user. So from the discussion my understanding is instead of copying the relevant files from the terraform adaptor, we are mounting a directory which will contain the required pre processing files. And we will be mounting the preprocessing files of all cloud providers and will execute the suitable pre-processing step based on the cloud-provider. Is this correct?

Reply:
If it is so then is it ok if we keep a directory with all the pre-processing steps required for each cloud provider in separate folders in the master node and copy the required directory contents to the terraform container as needed. Because from the wrapper function I can't find the cloud-provider unless I search the terraform template and like Jay said in AWS there might be cases where pre-processing is needed and pre-processing is not needed. So I can't just execute the preprocess file just because it exists in the case where pre-processing is not needed which happens when we do the volume mounting. If it needs to be done via volume mounting then we needed to process the tf.vars.json file from the wrapper function.

Reply:
Shouldn't the conditionals for checking if the pre/postprocessing script is needed reside inside the script itself? Rather than having the condition check in one place and the script logic in another? This should fit with the idea of mounting all scripts to the container instead of having to copy them as needed.

Reply:
"is it ok if we keep a directory with all the pre-processing steps required for each cloud provider in separate folders in the master node" -> YES. "copy the required directory contents to the terraform container as needed" -> NO. We should find a way to provide information for the wrapper function to make the decision whether to run preprocessing or not.

Reply:
@jaydesl I understood what you meant. Currently in the preprocess script that logic is there. I am checking the tfvars file for the preprocess flag which contains the cloud type. So if you mean we just need to check the flag to the required cloud provider and then call the required scripts for that I guess that is also good. There will only be one preprocess script which will act as an entry point for all the clouds and wrapper will call this preprocess script if it exists. The rest of the cloud specific script will be kept in a directory which will be mounted to terraform container. Is it correct?

Reply:
The file terraform/terraform wrapper executes only one piece of code which is called prechecks.py.
This file contains only egi related stuff. It is not clear to me how Jay's idea (about scanning the directory for other cloud-related preprocessor files written above) is implemented. Or did you drop this idea?

Reply:
@smith4 In the prechecks.py it will check which cloud providers pre-processing steps needs to be done based on the flag. Currently there is only egi. So if some other cloud providers with pre-processing steps are added, we will write how to call the pre-processing functions for that cloud provider. Like if the flag is aws, go to the folder for aws and execute the scripts in some particular order. That was the idea.

Reply:
@ResmiArjun I can see the flag check at the end of the prechecks.py file, but it still not a seemless solution for adding new clouds. When a new preprocess needs to be added, this prechecks.py must be updated. Jay suggested to have a general solution at the top where adding a new cloud does not require more than simply place an additional file inside the appropriate directory. Ok, I think we can proceed with the current solution if Jay agrees. I just did not understand the diff between Jay's idea and the implementation.

Reply:
Hi @smith4 / @ResmiArjun

Can the EGI-related functions inside _preprocess.py_ simply be moved to another EGI-specific module (eg. _egi_token.py_) and the utility functions be moved to a utils module (eg. _utils.py_)?

It would mean that preprocess.py does not grow exponentially if scripts for other clouds are added in the future

**utils.py**
```python
import json

def load_json():
    ...

def dump_json():
    ...
```


**egi_token.py**
```python
from utils import load_json, dump_json

tf_file = "/var/lib/micado/terraform/submitter/token.txt"
...

def update_token():
    ...

def preprocess_egi():
    ...
```

**preprocess.py**
```python
from egi_token import preprocess_egi
from utils import load_json, dump_json

def main():
    ...
    if cloud == "egi.eu":
        preprocess_egi(token_dict)
```